gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py

```diff
@@ -0,0 +1,107 @@
+"""Keypoint aggregator that de-duplicates keypoint detections, aggregating from image pairs to single images.
+
+Authors: John Lambert
+"""
+
+from typing import Dict, List, Optional, Tuple
+
+import numpy as np
+
+import gtsfm.utils.logger as logger_utils
+from gtsfm.common.keypoints import Keypoints
+from gtsfm.frontend.correspondence_generator.keypoint_aggregator.keypoint_aggregator_base import KeypointAggregatorBase
+
+logger = logger_utils.get_logger()
+
+
+class KeypointAggregatorDedup(KeypointAggregatorBase):
+    """Keypoint aggregator with de-duplication."""
+
+    def run(self, keypoints_dict: Dict[Tuple[int, int], Tuple[Keypoints, Keypoints]]) -> List[Optional[Keypoints]]:
+        """Aggregates per-pair image keypoints into a set of keypoints per image, with de-duplication.
+
+        Keypoints are computed per image pair, instead of per image, so they are aggregated per image here.
+
+        Args:
+            keypoints_dict: (i1,i2) maps to (keypoints_i1, keypoints_i2) representing matches (correspondences).
+
+        Returns:
+            keypoints_list: list of N Keypoints objects for N images.
+            putative_corr_idxs_dict: putative correspondence indices (K,2) for each image pair (i1,i2).
+        """
+        image_indices = set()
+        for (i1, i2) in keypoints_dict.keys():
+            image_indices.add(i1)
+            image_indices.add(i2)
+
+        max_img_idx = max(image_indices)
+        duplicates_found = 0
+
+        putative_corr_idxs_dict = {}
+        per_image_kpt_coordinates = {i: np.zeros((0, 2)) for i in image_indices}
+
+        # TODO(johnwlambert): use efficient algo (DSF) to find tracks once C++ variant implemented,
+        # instead of O(N^2) algo below.
+        # Have to merge keypoints across different views here (or turn off transitivity check).
+
+        for (i1, i2), (keypoints_i1, keypoints_i2) in keypoints_dict.items():
+            # both keypoints_i1 and keypoints_i2 have the same shape
+            N_to_add = keypoints_i1.coordinates.shape[0]
+
+            N1 = per_image_kpt_coordinates[i1].shape[0]
+            N2 = per_image_kpt_coordinates[i2].shape[0]
+
+            i1_count = N1
+            i1_indices = np.zeros(N_to_add)
+            unique_keypoints_i1_coordinates = []
+
+            for k1, uv1 in enumerate(keypoints_i1.coordinates):
+                is_identical1 = np.any(np.linalg.norm(per_image_kpt_coordinates[i1] - uv1, axis=1) == 0)
```

Comment:
I think you can do np.where(per_image_kpt_coordinates[i1] - uv1), the norm is unnecessary. You can also reuse that result below.

Reply:
Hmm I think it is needed to have the norm:
```python
>>> import numpy as np
>>> x = np.array([[1,2],[3,4],[5,6]])
>>> y = np.array([3,4])
>>> x - y
array([[-2, -2],
       [ 0,  0],
       [ 2,  2]])
>>> np.where(x-y)
(array([0, 0, 2, 2]), array([0, 1, 0, 1]))
>>> np.linalg.norm(x-y, axis=1)
array([2.82842712, 0.        , 2.82842712])
>>> np.where(np.linalg.norm(x-y, axis=1) == 0)
(array([1]),)
>>> 
```

Reply:
I'll update to re-use the computation though. Definitely wasteful to repeat it.

Reply:
Updated.

Reply:
np.where((x-y) == 0) ? should give x, y. It would give you coordinates of both columns in the row where the match is. If you just need one column, maybe a column-wise sum would help. 
norm is not cheap and should be avoided unless we need a distance - especially since this happens several times.